api/object-handlers.go

```diff
@@ -620,6 +567,83 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
 	pipeReader.Close()
 }
 
+// RenameObjectHandler - Rename Object
+// ----------
+//Do not support bucket to enable multiVersion renaming;
+//Folder renaming operation is not supported.
+func (api ObjectAPIHandlers) RenameObjectHandler(w http.ResponseWriter, r *http.Request){
+	helper.Debugln("RenameObjectHandler", "enter")
+	vars := mux.Vars(r)
+	BucketName := vars["bucket"]
+	targetObjectName := vars["object"]
+
+	//Determine if the renamed object is a folder
+	if hasSuffix(targetObjectName,"/") {
+		WriteErrorResponse(w, r, ErrInvalidRenameTarget)
+		return
+	}
+
+	var credential common.Credential
+	var err error
+	if credential, err = checkRequestAuth(api, r, policy.PutObjectAction, BucketName, targetObjectName); err != nil {
+		WriteErrorResponse(w, r, err)
+		return
+	}
+	sourceObjectName := r.Header.Get("X-Amz-Rename-Source-Key")
+
+	// X-Amz-Copy-Source should be URL-encoded
+	sourceObjectName, err = url.QueryUnescape(sourceObjectName)
+	if err != nil {
+		WriteErrorResponse(w, r, ErrInvalidRenameSourceKey)
+		return
+	}
+
+	//Determine if the renamed object is a folder
+	 if hasSuffix(sourceObjectName,"/"){
+		WriteErrorResponse(w, r, ErrInvalidRenameSourceKey)
+		return
+	 }
+
```

Comment:
What if the target object and the sourceObject has different prefix?

Reply:
With some improvements

Reply:
What we are dealing with is just the object, just to take into account the use of windows user folders, so the form of the folder is referenced on the client. In fact, the backend stores the Key-Value form, so it is only necessary to judge the possible error. Logic issues, the rest of the issues are supported by the client.